presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusMetadata.java

```diff
@@ -0,0 +1,179 @@
+
+        PrometheusTable table = prometheusClient.getTable(prometheusTableHandle.getSchemaName(), prometheusTableHandle.getTableName());
+        if (table == null) {
+            throw new TableNotFoundException(prometheusTableHandle.toSchemaTableName());
+        }
+
+        ImmutableMap.Builder<String, ColumnHandle> columnHandles = ImmutableMap.builder();
+        int index = 0;
+        for (ColumnMetadata column : table.getColumnsMetadata()) {
+            columnHandles.put(column.getName(), new PrometheusColumnHandle(column.getName(), column.getType(), index));
+            index++;
+        }
+        return columnHandles.build();
+    }
+
+    @Override
+    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        requireNonNull(prefix, "prefix is null");
+        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
+        for (SchemaTableName tableName : listTables(session, prefix)) {
+            ConnectorTableMetadata tableMetadata = getTableMetadata(tableName);
+            // table can disappear during listing operation
+            if (tableMetadata != null) {
+                columns.put(tableName, tableMetadata.getColumns());
+            }
+        }
+        return columns.build();
+    }
+
+    private ConnectorTableMetadata getTableMetadata(SchemaTableName tableName)
+    {
+        if (!listSchemaNames().contains(tableName.getSchemaName())) {
+            return null;
+        }
+
+        PrometheusTable table = prometheusClient.getTable(tableName.getSchemaName(), tableName.getTableName());
+        if (table == null) {
+            return null;
+        }
+
+        return new ConnectorTableMetadata(tableName, table.getColumnsMetadata());
+    }
+
+    private List<SchemaTableName> listTables(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        if (!prefix.getTable().isPresent()) {
+            return listTables(session, prefix.getSchema());
+        }
+        return ImmutableList.of(prefix.toSchemaTableName());
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        return ((PrometheusColumnHandle) columnHandle).getColumnMetadata();
+    }
+
+    @Override
+    public boolean usesLegacyTableLayouts()
+    {
+        return false;
+    }
+
+    @Override
+    public ConnectorTableProperties getTableProperties(ConnectorSession session, ConnectorTableHandle table)
+    {
+        return new ConnectorTableProperties();
+    }
+
+    @Override
+    public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
+    {
+        prometheusClient.setPredicate(Optional.of(constraint.getSummary()));
+        return Optional.empty();
```

Comment:
Optional.empty() means that predicate was not applied. You should store predicate in table handle you are returning and then use it for filtering when query prometheus.

Reply:
Done in b4e218c5